### ISSUE TITLE
503 Upstream connection failure issue with Envoy 1.13

### ISSUE THREAD

**Issue by shashank0791** [envoyproxy/envoy#28885]
We are observing intermittent Envoy 503 upstream connect failure issue when our client tries to stream data from Upstream server.

We are in the process of debugging this issue but so far we have 2 hypothesis:
1. Envoy race condition where it tries reusing connections that are closed while trying to communicate with upstream.
2. Client timeout is 5 seconds which is smaller than Envoy connection timeout (6 seconds). Client sends out RST after 5 seconds but Envoy never sees that (because of race condition) and still reuses same connection. 

So far, I have not been able to reproduce this issue in a controlled environment. We also do not have debug logs to help us RCA this issue.

Any help / pointers from Envoy community to debug further ?

**Comment by KBaichoo** [comment]
I would recommend looking at the access log output for the streams that are failing in this way: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage

To see the exact flags the access log reports. 

You might be able to figure out from https://github.com/envoyproxy/envoy/blob/e2eaff90d31ead71af55f1535e819a6b382e1f68/source/common/router/router.cc#L1343C4-L1346 what error message is getting mapped to `UpstreamConnectionFailure` if that's the flag occuring in your case.

You might also want to look at your timers: https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts.html in case the Envoy is timing out. 

cc @zuercher

**Comment by id-id-id** [comment]
We had the same problem.Are there any new discoveries?

**Comment by shashank0791** [comment]
We enabled debug mode. After searching through the logs, we found this behavior to be more prominent for streaming workloads where a connection could be consumed for a long time (until the streaming completes). To tackle this, we had to do a couple of things:

1. Retry on 503.
2. Increase connection pool size in Envoy.

we also were using an older version of Envoy (1.13) which had a race condition bug. Although this did not seem to be the only root cause of the 503 error as other non-streaming workloads were running just fine.
